### PR TITLE
[Feature] Admin Disable User

### PR DESCRIPTION
**Pull request by Ghalko**
## Proposal

Give the admin the ability to disable a user who requests that their account be disabled.

## Repercussions
Just the normal ones with power, it comes with responsibility.

## Deployment
Ensure that settings.ENABLE_EMAIL_SUBSCRIPTIONS = True for environments that actually have mailchimp API.

### CODE REVIEW

**Comment by sloria** [comment]
@whit537 Could you take a look this and Bryan's other admin PRs today if you get a chance? I will be booked fixing other issues today and likely won't be able to code review today, but I'd like to move some of these admin PRs forward.

**Comment by whit537** [comment]
Sure thing! :-)

**Comment by sloria** [comment]
Thanks Chad!

**Comment by whit537** [comment]
What's the Jira ticket for this?

**Comment by whit537** [comment]
Admin, huh? Haven't seen that yet, let's see what we've got ...

**Comment by Ghalko** [comment]
Hi @whit537, just to let you know that the admin module has its own set of invoke tasks, assets, and the like, just so you can get up and running. You will need the OSF mongo instance running in order to actually do anything with the admin if you go so far as to test this. It is ADM-4

Also much of this is based on #4960 which needs to be reviewed or merged.

**Comment by whit537** [comment]
Thanks @Ghalko, got distracted there for a bit, looking at this now ...

**Comment by whit537** [comment]
@Ghalko As a heads-up, my practice is to think out loud during reviews. You can chime in at any time, but you don't have to—I'm sure you have other things to work on. Once I finish a pass through, I'll have compiled a punchlist (here's a recent [example](https://github.com/pattisdr/osf.io/pull/2#issuecomment-184391751)), and then I'll ping you to put the :basketball: back in your court. :-)

**Comment by whit537** [comment]
Sooooooooo ... how do I stand up the admin? Let's check the README ...

**Comment by whit537** [comment]
Nada. [Docs](http://cosdev.readthedocs.org/)?

**Comment by whit537** [comment]
[Nope](http://cosdev.readthedocs.org/en/latest/search.html?q=admin&check_keywords=yes&area=default). To the source mobile! :blue_car:

**Comment by whit537** [comment]
I found [`inv adminserver`](https://github.com/CenterForOpenScience/osf.io/blob/e3527dd98f6733218d5c847c3ed25f0f60a5d47c/tasks/__init__.py#L81-L86), but am hitting an `ImportError`:

```
[osf.io]$ inv adminserver
DJANGO_SETTINGS_MODULE="admin.base.settings" python manage.py runserver 8001 --nothreading
[framework.mongo.handlers]  WARNING: Creating new client instance. 1 instances initialized.
Traceback (most recent call last):
  File "manage.py", line 7, in <module>
    from website.app import init_app
  File "/Users/whit537/workbench/cos/osf.io/website/app.py", line 18, in <module>
    from framework.transactions import handlers as transaction_handlers
  File "/Users/whit537/workbench/cos/osf.io/framework/transactions/handlers.py", line 9, in <module>
    from framework.transactions import utils, commands, messages
  File "/Users/whit537/workbench/cos/osf.io/framework/transactions/utils.py", line 8, in <module>
    from website.util import is_json_request
  File "/Users/whit537/workbench/cos/osf.io/website/util/__init__.py", line 18, in <module>
    from . import rubeus  # noqa
  File "/Users/whit537/workbench/cos/osf.io/website/util/rubeus.py", line 12, in <module>
    from framework.auth.decorators import Auth
  File "/Users/whit537/workbench/cos/osf.io/framework/auth/__init__.py", line 11, in <module>
    from .core import User, Auth
  File "/Users/whit537/workbench/cos/osf.io/framework/auth/core.py", line 35, in <module>
    from website import mails, settings, filters, security
  File "/Users/whit537/workbench/cos/osf.io/website/mails/__init__.py", line 1, in <module>
    from .mails import *   # noqa
  File "/Users/whit537/workbench/cos/osf.io/website/mails/mails.py", line 26, in <module>
    from framework.email import tasks
  File "/Users/whit537/workbench/cos/osf.io/framework/email/tasks.py", line 7, in <module>
    import sendgrid
ImportError: No module named sendgrid
[osf.io]$ 
```

Investigating ...

**Comment by whit537** [comment]
I don't see an admin-specific requirements [file](https://github.com/CenterForOpenScience/osf.io/tree/e3527dd98f6733218d5c847c3ed25f0f60a5d47c/requirements). I do see sendgrid in the main [`requirements.txt`](https://github.com/CenterForOpenScience/osf.io/blob/e3527dd98f6733218d5c847c3ed25f0f60a5d47c/requirements.txt). Is that new? Do I need to rerun pip install?

**Comment by whit537** [comment]
Blorg. Got a gcc failure ... in what?

**Comment by whit537** [comment]
gevent, looks like ...

**Comment by whit537** [comment]
https://github.com/gevent/gevent/issues/656

**Comment by whit537** [comment]
Alright, the workaround over there worked. I'm on to:

> You have unapplied migrations; your app may not work properly until they are applied.
Run 'python manage.py migrate' to apply them.

And:

screen shot 2016-02-19 at 10 12 38 pm

**Comment by whit537** [comment]
I can haz admin?

screen shot 2016-02-19 at 10 16 18 pm

**Comment by whit537** [comment]
> It is ADM-4

What does that mean, I wonder?

**Comment by whit537** [comment]
/me pokes around, finds `defaults.py` with a SQLite [configuration](https://github.com/CenterForOpenScience/osf.io/blob/e3527dd98f6733218d5c847c3ed25f0f60a5d47c/admin/base/settings/defaults.py#L113-L118) ...

**Comment by whit537** [comment]
> You will need the OSF mongo instance running in order to actually do anything with the admin

How does the SQLite config relate to that? What the heck is my login? :-)

**Comment by whit537** [comment]
```
[osf.io] $ inv admin.manage migrate
<snip>
[osf.io] $ inv admin.manage createsuperuser
<snip>
```

Yessssssssss.

screen shot 2016-02-19 at 10 39 13 pm

**Comment by whit537** [comment]
Okay! Pretty nice, straightforward admin. What is this PR about, then ... ?

**Comment by whit537** [comment]
> Give the admin the ability to disable a user who requests that their account be disabled.

**Comment by whit537** [comment]
Seems to do what it says on the tin.

screen shot 2016-02-19 at 10 44 26 pm
screen shot 2016-02-19 at 10 45 24 pm
screen shot 2016-02-19 at 10 44 52 pm

**Comment by whit537** [comment]
Reactivating also seems to work.

screen shot 2016-02-19 at 10 47 30 pm

**Comment by whit537** [comment]
@Ghalko Do you have any suggestions for how to review a PR like this, which includes unrelated commits? That's something I haven't done before. My usual practice is to use GitHub's "Files changed" tab on PRs, but that's useless here since it presents the entire diff with the base, and 70% (32 / 46) of the commits here are actually out of scope.

I suppose I could try generating a one-off compare view here on GitHub ... but I won't have inline PR diff comments then, will I?

**Comment by whit537** [comment]
I tried this comparison:

https://github.com/Ghalko/osf.io/compare/feature/admin_node...feature/admin_user_disable

But it has *more* commits than are on here, and there doesn't appear to be a straightforward relationship between #4951, #4960, and this PR—26% (18 / 68) of the commits there are merge commits.

**Comment by whit537** [comment]
Sorry, I got the base wrong: only 16% (18 / 111) of the commits are merge commits. :-)

**Comment by whit537** [comment]
> this and Bryan's other admin PRs

I find 13 [such](https://github.com/CenterForOpenScience/osf.io/pulls/Ghalko), three of which are tagged [WIP]. That seems like a lot! How do they all relate to each other?

**Comment by whit537** [comment]
The first commit on this PR is 0bffe7c2134519f73d3e17577a01c5dae3e68926. Looks like it is also present on seven other PRs.

screen shot 2016-02-19 at 11 25 20 pm

Clicking through confirms that those are all open, and that 0bffe7c2134519f73d3e17577a01c5dae3e68926 is the first commit on all of them, too. Sooooooo, which one does it *really* belong to? :-)

**Comment by whit537** [comment]
Well, #4951 is the oldest, and has the fewest commits. Seems like a likely candidate. ;)

**Comment by whit537** [comment]
#4951 is open and appears unreviewed. I'm going to turn my attention to that PR, and come back here once that and whatever else is landed first.

**Comment by Ghalko** [comment]
@whit537 sorry you were pretty much randomly assigned one of my numerous PRs that make up a bit of a backlog. Admin module isn't a priority until it will be. I have my own deadlines which are much more stringent, but your mention of no documentation makes me take pause. I'll probably build in some time this next sprint to address that. I will also give you a better map of what all of these depend on, but you won't be able to effectively use that tab until they also start getting merged.

**Comment by Ghalko** [comment]
Alright since you asked I will try to answer your questions, straighten out dependencies in one giant mess of a comment, following:

- The Jira ticket 'ADM-4' is on the admin project board. It is a separate project insofar as I am pretty exclusive to this project at the moment.
- (open two windows so I don't wear out my scroll wheel)
- Docs, I will be planning that in to this next sprint, or I am pretty much done with this one so I could probably bring those in. Also COS might need them in the very near future.
- PRs, with admin I have found that many small chunks are better, unfortunately it starts getting messy without some attention to CR and merging, hence why you were assigned, although to pretty much a random one. I really don't know why.
 - First the two that you should ignore: #4924 and #4976 The first is a PR that will make the OSF actually display the correct message to everyone on the OSF when someone reports spam. That is on hold until it becomes a priority. The second one is a hope, hopefully not just a dream to get Django up to 1.9, as there are many nice improvements done in the FormView class that I wouldn't have to just copy from Django source...
 - #4860 should be ready to merge, hoping that will happen soonish.
 - #4951 has been assigned to @cwisecarver for a while for CR, but he seems busy. I've been trying to stay with merges rather than mixing my merges and rebases.
 - #4960 This is the big one that most of the following are based on. Most of the following PRs have something to do with users or Nodes, so this is pretty much the base that most should use.
 - #5001 Simple addition of logs, I would really like to get this one merged reasonably quickly. Probably my highest priority after #4960 just because I want to have feedback before adding logging to everything. Refactor on small PR fine, refactor on 9 subsequent PRs, not so fun.
 - #5013 If anything this might replace #4960. It is a refactor of how the searches and displays are done for Users and Nodes. I feel like I am building the great wall of China, but I have a staging area where I build pieces and then they just sit for a long time. Maybe we'll find that staging area in future archeological digs.
 - #5020 is pretty much a throwaway PR, I wouldn't mind comments on how I inject information into a modal, because I start to use that quite often in other PRs. Even though this going to be replaced by desk integration (hopefully this next sprint) it is still an important PR.
 - #5040 Pretty much what you've seen. I believe this is currently up to date with #5013 
 - #5041 Another very similar PR to #5040 just with a different action for user administration
 - #5042 Another user admin action The reason these are split into separate PRs is to allow merging of more important features first.
 - #5045 is a request from community to sort based on the most recent reports of spam.
 - #5047 is a @sloria request for community team to be able to see registrations, very heavily based on work done for nodes in #5013.

I hope that clears up priorities and it might make it easier to CR just knowing the general idea of how these PRs are structured and their current languishment adjustment.

**Comment by cwisecarver** [comment]
@Ghalko Sorry, I remember having that IRL conversation about #4951 and ListViews. I didn't know you'd gone back and switched them to that. I'll make sure I take a look at the revised PR today or tomorrow.

**Comment by whit537** [comment]
> sorry you were pretty much randomly assigned one of my numerous PRs that make up a bit of a backlog.
> hence why you were assigned, although to pretty much a random one. I really don't know why.

No worries, I don't mind being thrown in the deep end. :-)

Thanks for the orientation. Seems like #4951 is a good place to start, then, ya?

**Comment by whit537** [comment]
... unless @cwisecarver gets to it first. I'm currently on #5033.

File: .gitignore [hunk 1]
```diff
@@ -114,4 +114,5 @@ admin/static/public/
 admin/webpack-stats.json
 # bower
 admin/static/vendor/
-admin/db.sqlite3
```
Comment (mfraezz) [inline .gitignore]: Why was this changed? It's a file for me, not a directory.

File: admin/users/views.py [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+
+from admin.base.views import GuidFormView, GuidView
+from admin.users.templatetags.user_extras import reverse_user
+from .serializers import serialize_user
+
+
+def disable_user(request, guid):
+    user = User.load(guid)
+    # user.disable_account()
+    user.is_disabled = True  # TODO: change to user.disable_account() after local tests
```
Comment (mfraezz) [inline admin/users/views.py]: This `TODO` should be done. Tests may require mocking some `mailchimp_utils`, I'm not sure if our test suite handles that.

File: admin/users/views.py [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+    user.save()
+    return redirect(reverse_user(guid))
+
+
+def reactivate_user(request, guid):
```
Comment (mfraezz) [inline admin/users/views.py]: This method also needs to resubscribe the user to `website.settings.MAILCHIMP_GENERAL_LIST` if `disable_user` is unsubscribing them (which it seems like it should). See https://github.com/CenterForOpenScience/osf.io/blob/develop/website/mailchimp_utils.py#L98-L101

**Comment by mfraezz** [comment]
Pass complete. Please also add tests and resolve merge conflicts :candy:

File: .gitignore (hunk 1 above)
Comment (Ghalko) [inline .gitignore]: Not sure why that was changed, changed back.

File: admin/users/views.py (hunk 2 above)
Comment (Ghalko) [inline admin/users/views.py]: @sloria I get a `ConnectionError: nodename nor servname provided, or not known` This isn't in the standard mailchimp error library like the other error that is caught and passed in the `disable_account` 

If I take out a fake api key '2-2' then I just get a straight-up `RuntimeError: An API key is required to connect to Mailchimp.` Not a very subtle error.

File: admin/users/views.py (hunk 3 above)
Comment (Ghalko) [inline admin/users/views.py]: Probably without the signal...

File: admin/base/views.py [hunk 4]
```diff
@@ -74,3 +57,15 @@ def form_valid(self, form):
...
+
+class GuidView(DetailView):
+    def __init__(self):
+        self.guid = None
+        super(GuidView, self).__init__()
+
+    def get(self, request, *args, **kwargs):
+        try:
+            return super(GuidView, self).get(request, *args, **kwargs)
+        except AttributeError:
```
Comment (sloria) [inline admin/base/views.py]: I think I've asked this in another PR, but why do you expect an `AttributeError` here?

File: admin/base/views.py [hunk 5]
```diff
@@ -74,3 +57,15 @@ def form_valid(self, form):
     @property
     def success_url(self):
         raise NotImplementedError
+
+
+class GuidView(DetailView):
+    def __init__(self):
+        self.guid = None
```
Comment (sloria) [inline admin/base/views.py]: Why is it necessary to save `guid` as an attribute of the view? I have a slightly bad feeling about overriding `__init__`; we should try to avoid overrides (except for hooks and methods that are meant to be overriden) if possible.

File: admin/nodes/views.py [hunk 6]
```diff
@@ -1,17 +1,23 @@
...
     def success_url(self):
         return reverse_node(self.guid)
+
+
+class NodeView(GuidView):
+    template_name = 'nodes/node.html'
+    context_object_name = 'node'
+
+    def get_object(self, queryset=None):
+        self.guid = self.kwargs.get('guid', None)
```
Comment (sloria) [inline admin/nodes/views.py]: As suggested in https://github.com/CenterForOpenScience/osf.io/pull/5040/files#r55427109, I don't think this is necessary.

File: admin/spam/urls.py [hunk 7]
```diff
@@ -4,6 +4,9 @@
 
 urlpatterns = [
     url(r'^$', views.SpamList.as_view(), name='spam'),
-    url(r'^(?P<spam_id>[a-z0-9]+)/$', views.SpamDetail.as_view(), name='detail'),
-    url(r'^(?P<spam_id>[a-z0-9]+)/email/$', views.email, name='email'),
+    url(r'^user/id-(?P<user_id>[a-z0-9]+)/$', views.UserSpamList.as_view(),
```
Comment (sloria) [inline admin/spam/urls.py]: The `id-` prefix is unnecessary. The URLs should look like `/user/abc12/`.

File: admin/spam/urls.py [hunk 8]
```diff
@@ -4,6 +4,9 @@
...
+        name='user_spam'),
+    url(r'^id-(?P<spam_id>[a-z0-9]+)/$', views.SpamDetail.as_view(),
```
Comment (sloria) [inline admin/spam/urls.py]: Same here: `id-` is unnecessary.

File: admin/spam/views.py [hunk 9]
```diff
@@ -37,47 +37,90 @@ def get_context_data(self, **kwargs):
...
             'page_number': page.number
         }
-        return super(SpamList, self).get_context_data(**context)
+
+
+class UserSpamList(SpamList):
+    template_name = 'users/spam.html'
+
+    def __init__(self):
+        self.user_id = None
```
Comment (sloria) [inline admin/spam/views.py]: Same here: avoid unnecessary overrides and attributed.

File: admin/spam/views.py [hunk 10]
```diff
@@ -37,47 +37,90 @@ def get_context_data(self, **kwargs):
...
+    def get_queryset(self):
+        self.status = self.request.GET.get('status', u'1')
+        self.user_id = self.kwargs.get('user_id', None)
+        query = (
+            Q('reports', 'ne', {}) &
+            Q('reports', 'ne', None) &
+            Q('user', 'eq', self.user_id) &
+            Q('spam_status', 'eq', int(self.status))
+        )
+        return Comment.find(query).sort(self.ordering)
```
Comment (sloria) [inline admin/spam/views.py]: Where does `self.ordering` get set?

File: admin/spam/views.py [hunk 11]
```diff
@@ -37,47 +37,90 @@ def get_context_data(self, **kwargs):
...
+
+    def get_context_data(self, **kwargs):
+        queryset = kwargs.pop('object_list', self.object_list)
+        page_size = self.get_paginate_by(queryset)
+        paginator, page, queryset, is_paginated = self.paginate_queryset(
+            queryset, page_size)
+        return {
+            'spam': map(serialize_comment, queryset),
```
Comment (sloria) [inline admin/spam/views.py]: Why not just pass the queryset and work `Comment` objects in the template?

```python
'spam': queryset,
```

```jinja
{% for comment in spam %}
...
comment.text
```

**Comment by sloria** [comment]
Pass finished.

File: admin/base/views.py (hunk 4 above)
Comment (Ghalko) [inline admin/base/views.py]: Yes, and its because when loading a guid if it comes back as a `None` then serializing it will return an attribute error which is raised by `get_object` which is then caught by `get` which is what actually returns the rendered web page, in this case a 404.

File: admin/base/views.py (hunk 5 above)
Comment (Ghalko) [inline admin/base/views.py]: Super is called the line below. The extra things for guids and similar have been taken care of in one PR but this and at least 11 others like it are waiting for that one to get merged. I believe #5013. @sloria

File: admin/nodes/views.py (hunk 6 above)
Comment (Ghalko) [inline admin/nodes/views.py]: Right. Will update this PR when others are merged before it. #5013

File: admin/spam/urls.py (hunk 7 above)
Comment (Ghalko) [inline admin/spam/urls.py]: Yes. As addressed in two or three other comments on PRs right now this will be fixed in a future PR where this can be taken care of all at once. @sloria @GaryKriebel Basically this was a bad decision that has propagated to a dozen PRs. To fix one way in one and a slightly different way in another will cause more merge conflicts then I am already handling (19+ files). This will be a simple fix when more PRs are merged and will not effect usage in the few days that it might actually get to be used.

File: admin/spam/urls.py (hunk 8 above)
Comment (Ghalko) [inline admin/spam/urls.py]: Yep. See above comment.

File: admin/spam/views.py (hunk 10 above)
Comment (Ghalko) [inline admin/spam/views.py]: It is a class attribute.

File: admin/spam/views.py (hunk 9 above)
Comment (Ghalko) [inline admin/spam/views.py]: Thanks, this may well be fixed in a yet unmerged PR. I don't remember, hard to keep track of in a dozen interrelated PRs. @sloria @GaryKriebel

File: admin/spam/views.py (hunk 11 above)
Comment (Ghalko) [inline admin/spam/views.py]: I think this is dealt with in either #5045 or #5013. Though #5013 is obviously higher priority.

**Comment by Ghalko** [comment]
Fixed default email subscriptions and associated tests. 

Overall the advice offered on this last review was addressing design decisions made about 3 sprints ago. I am aware of these issues and would dearly like to submit a PR fixing many of these not so great decisions, however, I am not completely sure which of the currently outstanding PRs might have fixed these, and if so if I should then merge that PR into all subsequent PRs. Basically untimely CRs have nearly completely wrecked this sprint and I hope to get some semblance of a process planned for this one. As it is @mfraezz has been bringing up many of these same issues and similar in his CRs. With his responsiveness I believe that PRs I commit could actually get timely CR and would avoid this very late, repetitive, and frustrating CR type for all parties involved. @sloria @GaryKriebel

**Comment by sloria** [comment]
@Ghalko OK, I see now #5013 fixes a number of the flaws in this PR; I didn't see that one before I reviewed this one. No need to backport those fixes to past PRs. I'll do my best to merge these older PRs which your other PRs are dependent on. Thanks for your patience.
